packages/assert/src/assert.js

```diff
@@ -57,6 +57,19 @@ export { quote as q };
 
 export { makeAssert };
 
+/**
+ * @template T
+ * @param {T | null | undefined} val
+ * @param {string} [message]
+ * @returns {T}
+ */
+export function assertDefined(val, message) {
+  if (val != null) {
+    return val;
+  }
+  throw Error(message !== undefined ? message : `unexpected ${val}`);
+}
```

Comment:
I like the idea of it, but have some questions. The assert module is migrating from agoric-sdk to endo anyway. But wherever it is, it should be written in our normal `assertFoo` style. The following is how I might write it.

Also, please don't write `function` functions. We have essentially adopted an arrow-only style. We have not yet taken the time to convert old code. But new code should be in arrow-only style unless there's a good stated reason otherwise.

Also, you're exporting a non-hardened value.

```suggestion
/**
 * @template T
 * @param {T | null | undefined} val
 * @param {string} [optDetails]
 * @returns {T}
 */
export const assertDefined = (val, optDetails = X`unexpected ${q(val)}`) => {
  if (val != null) {
    // This `!= null` idiom checks that `val` is not `null` nor `undefined`.
    return val;
  }
  assert.fail(optDetails);
};
harden(assertDefined);
```

Also, I'm not sure I like "defined" as the opposite of "nullish". But I cannot think of a better one.

Reply:
Apropos naming, is there a better term than "assert"? In SES, [global assert](https://github.com/endojs/endo/blob/master/packages/ses/index.d.ts#L168-L195) mostly return void but this one returns the argument.

I could see someone being surprised by one or the other. For consistency, would it be better to return the argument when asserting a type or to have a different verb for this version that returns its argument? cc @kriskowal

Reply:
This is a good point. I agree that we should not have a function named `assertFoo` that returns its argument. This is indeed inconsistent with both precedents: The `assertFoo` functions we have don't return their argument, and the validating or coercing functions we have that do return their argument, or a coerced-to-validated form of their argument, like `Nat`, and written simply naming the property they assure about their return result. This one at least is written with an initial upper case, mirroring similar usage of `String`, `BigInt` etc.

How about simply `NotNullish` ?

Reply:
Went with slight tweak, `NonNullish`, similar to https://www.typescriptlang.org/docs/handbook/utility-types.html#nonnullabletype

Reply:
I have in hobby projects found it useful to distinguish `assert*` (returns void) from `assume*` (returns argument), though one could argue that all `assume` functions could just be `assert` and ignore the return.

https://github.com/borkshop/js/blob/main/cube/assert.js

Not having to name an intermediate variable in an expression like this is useful:

```js
const location = assumeDefined(locations.get(entityId), X`No entity for identifier ${q(entityId)}`);
```

Reply:
With some refactoring I don't need this anymore but I still think it would be useful. Pulling out to https://github.com/Agoric/agoric-sdk/pull/4422 and removing you from reviewers for this PR.